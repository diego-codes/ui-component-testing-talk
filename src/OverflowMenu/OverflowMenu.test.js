import React from 'react';
import { render } from 'react-testing-library';
import { OverflowMenu, OverflowMenuItem } from '.';

describe(OverflowMenu.name, () => {
  describe('Structure', () => {
    it('renders closed', () => {
      const { container } = render(
        <OverflowMenu toggleText="Toggle menu">
          <OverflowMenuItem>Option 1</OverflowMenuItem>
          <OverflowMenuItem>Option 2</OverflowMenuItem>
        </OverflowMenu>
      );
      expect(container).toMatchSnapshot();
    });
  });

  describe('Behavior', () => {
    it('renders options when opened', () => {
      const { container, getByTestId } = render(
        <OverflowMenu toggleText="Toggle menu">
          <OverflowMenuItem>Option 1</OverflowMenuItem>
          <OverflowMenuItem>Option 2</OverflowMenuItem>
        </OverflowMenu>
      );
      getByTestId('overflow-menu-toggle').click();
      expect(container).toMatchSnapshot();
    });

    it('hides options when toggle button is pressed twice', () => {
      const { container, getByTestId } = render(
        <OverflowMenu toggleText="Toggle menu">
          <OverflowMenuItem>Option 1</OverflowMenuItem>
          <OverflowMenuItem>Option 2</OverflowMenuItem>
        </OverflowMenu>
      );
      getByTestId('overflow-menu-toggle').click();
      getByTestId('overflow-menu-toggle').click();
      expect(container).toMatchSnapshot();
    });

    it('closes menu when user clicks outside of menu', () => {
      const { container, getByTestId } = render(
        <OverflowMenu toggleText="Toggle menu">
          <OverflowMenuItem>Option 1</OverflowMenuItem>
          <OverflowMenuItem>Option 2</OverflowMenuItem>
        </OverflowMenu>
      );
      getByTestId('overflow-menu-toggle').click();
      document.body.click();
      expect(container).toMatchSnapshot();
    });

    it('does not close menu when user clicks inside of menu', () => {
      const { container, getByTestId } = render(
        <OverflowMenu toggleText="Toggle menu">
          <OverflowMenuItem data-testid="option">Option 1</OverflowMenuItem>
          <OverflowMenuItem>Option 2</OverflowMenuItem>
        </OverflowMenu>
      );
      getByTestId('overflow-menu-toggle').click();
      getByTestId('option').click();
      expect(container).toMatchSnapshot();
    });

    it('removes click event listener when component unmounts', () => {
      const { unmount } = render(
        <OverflowMenu toggleText="Toggle menu">
          <OverflowMenuItem>Option 1</OverflowMenuItem>
          <OverflowMenuItem>Option 2</OverflowMenuItem>
        </OverflowMenu>
      );
      const windowRemoveEvenListenerSpy = jest.spyOn(window, 'removeEventListener');
      unmount();
      expect(windowRemoveEvenListenerSpy).toHaveBeenCalled();
    });
  });
});
