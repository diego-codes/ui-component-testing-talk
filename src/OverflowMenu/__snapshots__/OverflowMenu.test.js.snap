// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`OverflowMenu Behavior closes menu when user clicks outside of menu 1`] = `
<div>
  <div
    class="sc-bdVaJa huqire"
    data-testid="overflow-menu"
  >
    <button
      class="sc-bwzfXH sc-htpNat gTrUOb"
      data-testid="overflow-menu-toggle"
    >
      Toggle menu
    </button>
  </div>
</div>
`;

exports[`OverflowMenu Behavior does not close menu when user clicks inside of menu 1`] = `
<div>
  <div
    class="sc-bdVaJa huqire"
    data-testid="overflow-menu"
  >
    <button
      class="sc-bwzfXH sc-htpNat tfKyv"
      data-testid="overflow-menu-toggle"
      open=""
    >
      Toggle menu
    </button>
    <div
      class="sc-ifAKCX dEJfdm"
    >
      <button
        class="sc-bwzfXH sc-bxivhb cCHJQU"
        data-testid="option"
      >
        Option 1
      </button>
      <button
        class="sc-bwzfXH sc-bxivhb cCHJQU"
      >
        Option 2
      </button>
    </div>
  </div>
</div>
`;

exports[`OverflowMenu Behavior hides options when toggle button is pressed twice 1`] = `
<div>
  <div
    class="sc-bdVaJa huqire"
    data-testid="overflow-menu"
  >
    <button
      class="sc-bwzfXH sc-htpNat gTrUOb"
      data-testid="overflow-menu-toggle"
    >
      Toggle menu
    </button>
  </div>
</div>
`;

exports[`OverflowMenu Behavior renders options when opened 1`] = `
<div>
  <div
    class="sc-bdVaJa huqire"
    data-testid="overflow-menu"
  >
    <button
      class="sc-bwzfXH sc-htpNat tfKyv"
      data-testid="overflow-menu-toggle"
      open=""
    >
      Toggle menu
    </button>
    <div
      class="sc-ifAKCX dEJfdm"
    >
      <button
        class="sc-bwzfXH sc-bxivhb cCHJQU"
      >
        Option 1
      </button>
      <button
        class="sc-bwzfXH sc-bxivhb cCHJQU"
      >
        Option 2
      </button>
    </div>
  </div>
</div>
`;

exports[`OverflowMenu Structure renders closed 1`] = `
<div>
  <div
    class="sc-bdVaJa huqire"
    data-testid="overflow-menu"
  >
    <button
      class="sc-bwzfXH sc-htpNat gTrUOb"
      data-testid="overflow-menu-toggle"
    >
      Toggle menu
    </button>
  </div>
</div>
`;
